delira/training/experiment.py

```diff
@@ -1,388 +1,583 @@
+    Baseclass for Experiments.
+
+    Implements:
+
+    * Setup-Behavior for Models, Trainers and Predictors (depending on train
+        and test case)
 
-    See Also
-    --------
-    :class:`PyTorchExperiment`
+    * The K-Fold logic (including stratified and random splitting)
+
+    * Argument Handling
 
     """
-    @abstractmethod
-    def __init__(self, n_epochs, *args, **kwargs):
+
+    def __init__(self,
+                 params: Parameters,
+                 model_cls: AbstractNetwork,
+                 n_epochs=None,
+                 name=None,
+                 save_path=None,
+                 key_mapping=None,
+                 val_score_key=None,
+                 optim_builder=None,
+                 checkpoint_freq=1,
+                 trainer_cls=BaseNetworkTrainer,
+                 **kwargs):
         """
 
         Parameters
         ----------
-        n_epochs : int
-            number of epochs to train
-        *args :
-            positional arguments
+        params : :class:`Parameters`
+            the training parameters
+        model_cls : Subclass of :class:`AbstractNetwork`
+            the class implementing the model to train
+        n_epochs : int or None
+            the number of epochs to train, if None: can be specified later
+            during actual training
+        name : str or None
+            the Experiment's name
+        save_path : str or None
+            the path to save the results and checkpoints to.
+            if None: Current working directory will be used
+        key_mapping : dict
+            mapping between data_dict and model inputs (necessary for
+            prediction with :class:`Predictor`-API)
+        val_score_key : str or None
+            key defining which metric to use for validation (determining best
+            model and scheduling lr); if None: No validation-based operations
+            will be done (model might still get validated, but validation metrics
+            can only be logged and not used further)
+        optim_builder : function
+            Function returning a dict of backend-specific optimizers
+        checkpoint_freq : int
+            frequency of saving checkpoints (1 denotes saving every epoch,
+            2 denotes saving every second epoch etc.); default: 1
+        trainer_cls : subclass of :class:`BaseNetworkTrainer`
+            the trainer class to use for training the model
         **kwargs :
-            keyword arguments
+            additional keyword arguments
+
         """
+
+        if n_epochs is None:
+            n_epochs = params.nested_get("n_epochs",
+                                         params.nested_get("num_epochs"))
+
         super().__init__(n_epochs)
-        self._run = 0
 
-    @abstractmethod
-    def setup(self, *args, **kwargs):
-        """
-        Abstract Method to setup a :class:`AbstractNetworkTrainer`
+        # params could also be a file containing a pickled instance of parameters
```

Comment:
if we accept filepaths for params, this must come before handling `n_epochs is None`